pkg/common/cns-lib/volume/taskmanager.go

```diff
@@ -0,0 +1,355 @@
+	listView := t.getListView()
+	if listView == nil {
+		log.Errorf("ListView is empty. Will attempt to remove invalid tasks in next attempt. ")
+		return
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	log.Infof("pending tasks count before purging: %v", len(t.taskToChannelMap))
+	for taskDetails := range t.taskToChannelMap {
+		if taskDetails.MarkedForRemoval == true {
+			err := listView.Remove(ctx, []types.ManagedObjectReference{taskDetails.Reference})
+			if err != nil {
+				log.Errorf("failed to remove task from ListView. error: %+v", err)
+				continue
+			}
+			delete(t.taskToChannelMap, taskDetails)
+		}
+	}
+	log.Infof("pending tasks count after purging: %v", len(t.taskToChannelMap))
+}
+
+// MarkForDeletion marks a given task MoRef for deletion by setting a boolean flag in the TaskDetails object
+func (t *TaskViewManagerImpl) MarkForDeletion(ctx context.Context, taskMoRef types.ManagedObjectReference) error {
+	log := logger.GetLogger(ctx)
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	ch, ok := t.taskToChannelMap[TaskDetails{Reference: taskMoRef}]
+	if !ok {
+		return logger.LogNewErrorf(log, "failed to retrieve receiver channel for task %+v", taskMoRef)
+	}
+	delete(t.taskToChannelMap, TaskDetails{Reference: taskMoRef})
+	t.taskToChannelMap[TaskDetails{Reference: taskMoRef, MarkedForRemoval: true}] = ch
+	return nil
+}
+
+func (t *TaskViewManagerImpl) GetPendingTasksCount() int {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	return len(t.taskToChannelMap)
+}
+
+func NewTaskViewManager(virtualCenter *cnsvsphere.VirtualCenter) *TaskViewManagerImpl {
+	ctx := context.Background()
+	m := make(map[TaskDetails]chan Result)
+	t := &TaskViewManagerImpl{
+		virtualCenter:    virtualCenter,
+		ctx:              ctx,
+		taskToChannelMap: m,
+	}
+	go t.init(t.ctx)
+	return t
+}
+
+// TaskViewManagerImpl is the struct used to manage a single listView instance.
+// params
+// virtualCenter: holds a reference to the global VC object
+// sync.Mutex: is required to handle race conditions for
+// storing and retrieving tasks during a driver restart
+// taskToChannelMap: is an in-memory map to associate task details
+// to the channel created by the caller to receive the task result
+// context.Context: is separate from the context set by CSI ops
+// listView: holds the managed object used to monitor multiple concurrent VC tasks
+// shouldStopListening: in case of regular CSI operation, even after receiving a batch of updates,
+// we want to continue listening for subsequent updates.
+// in case of unit tests, we need to stop listening for the test to complete execution
+type TaskViewManagerImpl struct {
+	mu                  sync.Mutex
+	virtualCenter       *cnsvsphere.VirtualCenter
+	listView            *view.ListView
+	ctx                 context.Context
+	taskToChannelMap    map[TaskDetails]chan Result
+	shouldStopListening bool
+}
+
+func (t *TaskViewManagerImpl) ResetVirtualCenter(virtualCenter *cnsvsphere.VirtualCenter) {
```

Comment:
A bit deceiving imo, the code just sets the VirtualCenter to whatever was passed in as param.

Reply:
Renamed to SetVirtualCenter()